lib/src/sql/parser.rs

```diff
@@ -1,87 +1,30 @@
 use crate::err::Error;
 use crate::sql::error::Error::ParserError;
+use crate::sql::error::IResult;
 use crate::sql::query::{query, Query};
 use crate::sql::thing::Thing;
 use crate::sql::value::Value;
 use nom::Err;
 use std::str;
 
 pub fn parse(input: &str) -> Result<Query, Error> {
-	match input.trim().len() {
-		0 => Err(Error::QueryEmpty),
-		_ => match query(input) {
-			Ok((_, query)) => Ok(query),
-			Err(Err::Error(e)) => match e {
-				ParserError(e) => {
-					let (s, l, c) = locate(input, e);
-					Err(Error::InvalidQuery {
-						line: l,
-						char: c,
-						sql: s.to_string(),
-					})
-				}
-			},
-			Err(Err::Failure(e)) => match e {
-				ParserError(e) => {
-					let (s, l, c) = locate(input, e);
-					Err(Error::InvalidQuery {
-						line: l,
-						char: c,
-						sql: s.to_string(),
-					})
-				}
-			},
-			_ => unreachable!(),
-		},
-	}
+	parse_impl(input, query)
 }
 
 pub fn thing(input: &str) -> Result<Thing, Error> {
-	match input.trim().len() {
-		0 => Err(Error::QueryEmpty),
-		_ => match super::thing::thing(input) {
-			Ok((_, query)) => Ok(query),
-			Err(Err::Error(e)) => match e {
-				ParserError(e) => {
-					let (s, l, c) = locate(input, e);
-					Err(Error::InvalidQuery {
-						line: l,
-						char: c,
-						sql: s.to_string(),
-					})
-				}
-			},
-			Err(Err::Failure(e)) => match e {
-				ParserError(e) => {
-					let (s, l, c) = locate(input, e);
-					Err(Error::InvalidQuery {
-						line: l,
-						char: c,
-						sql: s.to_string(),
-					})
-				}
-			},
-			_ => unreachable!(),
-		},
-	}
+	parse_impl(input, super::thing::thing)
 }
 
 pub fn json(input: &str) -> Result<Value, Error> {
+	parse_impl(input, super::value::json)
+}
+
+fn parse_impl<O>(input: &str, parser: impl Fn(&str) -> IResult<&str, O>) -> Result<O, Error> {
 	match input.trim().len() {
 		0 => Err(Error::QueryEmpty),
-		_ => match super::value::json(input) {
-			Ok((_, query)) => Ok(query),
-			Err(Err::Error(e)) => match e {
-				ParserError(e) => {
-					let (s, l, c) = locate(input, e);
-					Err(Error::InvalidQuery {
-						line: l,
-						char: c,
-						sql: s.to_string(),
-					})
-				}
-			},
-			Err(Err::Failure(e)) => match e {
+		_ => match parser(input) {
+			Ok((_, parsed)) => Ok(parsed),
+			Err(Err::Error(e)) | Err(Err::Failure(e)) => match e {
 				ParserError(e) => {
 					let (s, l, c) = locate(input, e);
 					Err(Error::InvalidQuery {
```

Comment:
This makes it easier to fix #1227 without code duplication (fix coming in future PR)